src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1548,56 +1548,92 @@ void RaftPart::processAppendLogRequest(
+            it.reset();
+        }
+
+        //  stale log
+        if (diffIndex == numLogs) {
+            resp.set_last_log_id(firstId + numLogs - 1);
+            resp.set_last_log_term(req.get_log_term());
+            resp.set_error_code(cpp2::ErrorCode::E_LOG_STALE);
+            //nothing to append
+            return;
+        }
+
+        //rollback the wal_
+        if (wal_->rollbackToLog(firstId + diffIndex -1)) {
             lastLogId_ = wal_->lastLogId();
             lastLogTerm_ = wal_->lastLogTerm();
-            resp.set_last_log_id(lastLogId_);
-            resp.set_last_log_term(lastLogTerm_);
             LOG(INFO) << idStr_ << "Rollback succeeded! lastLogId is " << lastLogId_
                       << ", logLogTerm is " << lastLogTerm_
                       << ", committedLogId is " << committedLogId_
                       << ", term is " << term_;
-         }
-         resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
-         return;
-    } else if (req.get_last_log_id_sent() > lastLogId_) {
-        // There is a gap
-        LOG(INFO) << idStr_ << "Local is missing logs from id "
-                << lastLogId_ << ". Need to catch up";
-        resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
-        return;
-    } else if (req.get_last_log_id_sent() < lastLogId_) {
-        // TODO(doodle): This is a potential bug which would cause data not in consensus. In most
-        // case, we would hit this path when leader append logs to follower and timeout (leader
-        // would set lastLogIdSent_ = logIdToSend_ - 1 in Host). **But follower actually received
-        // it successfully**. Which will explain when leader retry to append these logs, the LOG
-        // belows is printed, and lastLogId_ == req.get_last_log_id_sent() + 1 in the LOG.
-        //
-        // In fact we should always rollback to req.get_last_log_id_sent(), and append the logs from
-        // leader (we can't make promise that the logs in range [req.get_last_log_id_sent() + 1,
-        // lastLogId_] is same with follower). However, this makes no difference in the above case.
-        LOG(INFO) << idStr_ << "Stale log! Local lastLogId " << lastLogId_
-                  << ", lastLogTerm " << lastLogTerm_
-                  << ", lastLogIdSent " << req.get_last_log_id_sent()
-                  << ", lastLogTermSent " << req.get_last_log_term_sent();
-        resp.set_error_code(cpp2::ErrorCode::E_LOG_STALE);
-        return;
+        } else {
+            LOG(ERROR) << idStr_ << "Rollback fail! lastLogId is" << lastLogId_
+                       << ", logLogTerm is " << lastLogTerm_
+                       << ", committedLogId is " << committedLogId_
+                       << ", rollback id  is " << firstId + diffIndex -1;
+            resp.set_error_code(cpp2::ErrorCode::E_WAL_FAIL);
+            return;
+        }
+
+        // update msg
+        logEntries = std::vector<cpp2::LogEntry> (req.get_log_str_list().begin() + diffIndex,
```

Comment:
This could be quite expensive as well, emm, can't figure out a better way for now.

Reply:
I have refined it to avoid copy